### PR TITLE
JER integration into conditions database

### PR DESCRIPTION
**Pull request by blinkseb**
(This a a resubmission of #10350 targeting CMSSW 7.6.x and with corrections)

This PR adds the necessary machinery to store Jet energy Resolutions and associated Scale Factors to the conditions database.

Values are parsed from text files using the same format as Jet Energy Corrections text files. Two new records are added, JetResolutionRcd and JetResolutionScaleFactorRcd, storing respectively resolutions and scale factors. Both resolutions and scale factors are converted to a common object, JetResolutionObject, which is serialized into the corresponding records.

Facility are provided to create a database from text files [1] and to create text files from a database or a GT [2].

Users can access values using the classes JetResolution and JetResolutionScaleFactors. A static function is provided to hide the boilerplate to retrieve the object from the condition database. A demo analyzer is included as an how-to tutorial [3] and is ran as an unit-test for the package.

This depends also on two data files used for unit-testing. ~~Corresponding PR is pending for inclusion on the cms-data org [4]~~. PR is merged.

This interface has been presented and validated at the JERC meeting on 07/23.

This will need to be backported into CMSSW 7.5.x and probably also 7.4.x

[1] https://github.com/blinkseb/cmssw/blob/665bd5d702f5ed976a898b7f8c426a5955eaf103/JetMETCorrections/Modules/test/JetResolutionDBWriter_cfg.py
[2] https://github.com/blinkseb/cmssw/blob/665bd5d702f5ed976a898b7f8c426a5955eaf103/JetMETCorrections/Modules/test/JetResolutionLocalDBReader_cfg.py
[3] https://github.com/blinkseb/cmssw/blob/665bd5d702f5ed976a898b7f8c426a5955eaf103/JetMETCorrections/Modules/plugins/JetResolutionDemo.cc
[4] https://github.com/cms-sw/cmssw/pull/10350

### CODE REVIEW

**Comment by blinkseb** [comment]
Is there any news on this? This is quite high priority for JME. Thanks!

The PR for data has been merged on cms-data so everything is available in latest IB.

**Comment by davidlange6** [comment]
@cmsbuild, please test

**Comment by diguida** [comment]
@blinkseb sorry for late reply, looking into it.

File: CondFormats/JetMETObjects/interface/JetResolutionObject.h [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+            // Advanced usage
+            const JetResolutionObject* getResolutionObject() const {
+                return m_object.get();
+            }
+
+        private:
+            std::shared_ptr<JetResolutionObject> m_object;
+    };
+
+    class JetResolutionScaleFactor {
```
Comment (diguida) [inline CondFormats/JetMETObjects/interface/JetResolutionObject.h]: @blinkseb I am a bit confused by this class. It is basically an observer for `JetResolutionObject`.
I would definitively not put it here, where you declare entity classes (observers fall in the control class domain).
See more comments on the implementation on each method.

Comment (blinkseb) [inline CondFormats/JetMETObjects/interface/JetResolutionObject.h]: Sorry for the delay, I was in holidays.

Do you want me to move these classes into another files?

**Comment by blinkseb** [comment]
As suggested by @diguida, I moved the classes ``JetResolution`` and ``JetResolutionScaleFactors`` into ``JetMETCorrections/Modules``.

**Comment by schoef** [comment]
@diguida Could we please proceed with this? Anything wlese required from our side?

**Comment by mmusich** [comment]
please test

**Comment by blinkseb** [comment]
I've updated the PR to fix build issues.

**Comment by mmusich** [comment]
please test

**Comment by blinkseb** [comment]
Is there something else I need to do for this PR to be merged? Just let me know.

**Comment by mmusich** [comment]
@blinkseb we are reviewing it. Thanks

**Comment by diguida** [comment]
+1
* New condition object `JetResolutionObject`
* Import and fetch macros able to manage the object
* Two records associated to it
* Two control classes defined for later use in client code.
Things to be fixed or possibly reviewed in later PR:
* remove spurious include in https://github.com/cms-sw/cmssw/pull/10448/files#diff-69cc9f239fd56430a5acddec1b1fa1f3R5 as the control class `JetResolutionObject` which is handling the record via the `EventSetup` is now in a different package.
* check how you dispatch the `JetResolutionObject` payloads as in https://github.com/cms-sw/cmssw/pull/10448/files#diff-4069f7250ad296887d167d001ac989f9R23 and https://github.com/cms-sw/cmssw/pull/10448/files#diff-4069f7250ad296887d167d001ac989f9R48 (where you return by value the payload from the `EventSetup` delivered via `Records`)

**Comment by blinkseb** [comment]
Thanks for the review @diguida! I'll update the PR with the changes requested. Just one question about your latest point (dispatching of ``JetResolutionObject``).

I tried to return a const ref but it's obviously not possible since the ``Handle`` is local to the function. What would you advise to return? Directly the ``Handle`` maybe?

**Comment by diguida** [comment]
@ggovi ping :-)

**Comment by diguida** [comment]
@blinkseb do not update this PR now, otherwise it has to be re-tested!
Concerning your question, the issue is how to manage objects delivered via `EventSetup`. The quick reply is: do not use them outside the context where the `EventSetup` makes them available to the clients.
I discussed this with @ggovi , and we agreed on this "quick summary".
I hope @Dr15Jones can confirm this is correct.

**Comment by ggovi** [comment]
+1

**Comment by Dr15Jones** [comment]
Seems reasonable

**Comment by diguida** [comment]
@monttj @vadler ping :-)

**Comment by diguida** [comment]
@Dr15Jones thanks for looking into that.

**Comment by monttj** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**